### DISCUSSION TITLE
Modernize packaging?

### DISCUSSION THREAD

**Discussion by renefritze** [pymor/pymor#1869] [Ideas]
In the run up to the 2022.2 release someone asked if we have and/or actually (still) need a `setup.py`. 
My answer was something like "no, but we really don't want to touch that before the release".  
Now the question to everybody (@pymor/all) is: should I touch this after the release? There really is no pressure to. The current setup will (probably) continue to work until setuptools dies. But it duplicates information from `dependencies.py` into `setup.py`, `requirements*` and `pyproject.toml`. The modern approach would be getting rid of everything but `pyproject.toml` (maybe even `setup.cfg`) and switch the build over to `hatchling`.
On the one hand the simplification would be good future proofing, OTOH it's not really a need. What do you guys think?

**Comment by renefritze** [comment]
A softer version would be keeping setuptools, but removing `setup.py` like I did [here](https://github.com/renefritze/pypi_oldest_requirements/pull/10/files)
Also gets rid of versioneer and setup.cfg.

**Reply by artpelling** [reply]
I like it. Do you have an idea what to do with `flake8` then? It does not support `pyproject.toml`, right?
We could use [pyproject-flake8](https://github.com/csachs/pyproject-flake8) or move the config to a `.flake8` file..

**Reply by artpelling** [reply]
Also do you have any experience/opinion on [ruff](https://github.com/charliermarsh/ruff)?

**Reply by renefritze** [reply]
I wasn't aware that `flake8` refuses to support `pyproject.toml`. Thanks for pointing that out. If `flake8` would turn out to the only config blocking the removal of `setup.cfg` I would prefer the extra config file over the monkeypatch project. 
I have not used ruff yet. I do think that it's a very good alternative and we probably want to use it as a pre-commit hook at some point even.

**Reply by artpelling** [reply]
I just gave ruff a spin and it is so much faster as a pre-commit hook. It seems like we could also get rid of a lot of error ignores, the additional flake8 docstring modules and could even activate more things such as sorted imports and quote checking.

**Reply by renefritze** [reply]
Sounds great. Would you want to make a PR for the switch?

**Reply by artpelling** [reply]
Yes I could look into it. However, we will run into tons of new linting errors...

**Reply by renefritze** [reply]
We probably wouldn't want to enable all of ruff's checks at first then?

**Reply by artpelling** [reply]
Thats right. It has different modules that can be activated, similar to the the flake8 modules. Another thing I noticed is that it does not check for some error codes (e.g. `E128 continuation line under-indented for visual indent`).
